pkg/algorithm/cstorpoolselect/v1alpha1/select.go

```diff
@@ -225,53 +459,68 @@ func (s *selection) validate() error {
 // filter returns the final list of pools that
 // gets selected, after passing the original list
 // of pools through the registered selection policies
-func (s *selection) filter() ([]string, error) {
+func (s *selection) filter() (*csp.CSPList, error) {
 	var (
-		filtered []string
-		err      error
+		err error
 	)
-	if len(s.policies) == 0 {
-		return s.poolUIDs, nil
+	if len(s.policies.items) == 0 {
+		return s.pools, nil
 	}
 	// make a copy of original pool UIDs
-	filtered = append(filtered, s.poolUIDs...)
-	for _, policy := range s.policies {
+	filtered := &csp.CSPList{}
+	for _, csp := range s.pools.Items {
+		filtered.Items = append(filtered.Items, csp)
+	}
+
+	// Sorting policies based on the priority
+	policies := s.policies.sortByPriority()
+	// Executing policy filters
+	for _, policy := range policies {
 		filtered, err = policy.filter(filtered)
 		if err != nil {
 			return nil, err
 		}
+		// stopping here if running as
+		// singleExecution mode
+		if s.mode == singleExection {
+			break
+		}
 	}
 	return filtered, nil
 }
 
-// FilterWithBuildOptions will return filtered pool UIDs
-// from the provided list based on pool
-// selection options
-func FilterWithBuildOptions(origPoolUIDs []string, opts []buildOption) ([]string, error) {
-	return Filter(origPoolUIDs, opts...)
-}
-
-// Filter will return filtered pool UIDs
-// from the provided list based on pool
-// selection options
-func Filter(origPoolUIDs []string, opts ...buildOption) ([]string, error) {
-	if len(opts) == 0 {
-		return origPoolUIDs, nil
+// Filter will filter the provided pools
+// based on pool selection policies
+func Filter(originalPools *csp.CSPList, opts ...buildOption) (*csp.CSPList, error) {
+	if originalPools == nil {
+		return originalPools, nil
 	}
-	s := newSelection(origPoolUIDs, opts...)
+	s := newSelection(originalPools, singleExection, opts...)
```

Comment:
Please remove `singleExecution` as an argument. It will be part of `opts` & if not provided will default to single mode of execution. Defaulting behaviour is handled in the constructor of selection struct.

Reply:
removed